lib/clients/AliClient.py

```diff
@@ -69,4 +154,397 @@ def _download_from_blobstore(self, blob_to_download_name, blob_download_target_p
+            if snapshot.status == 'accomplished':
+                self._add_snapshot(snapshot.id)
+                self.output_json['snapshotId'] = snapshot.id
+                self.logger.info('{} SUCCESS: snapshot-id={}, volume-id={}, status={} with tags {}'.format(
+                    log_prefix, snapshot.id, volume_id, snapshot.status, self.tags))
+            else:
+                message = '{} ERROR: snapshot-id={} status={}'.format(
+                    log_prefix, snapshot_name, snapshot.status)
+                raise Exception(message)
+        except Exception as error:
+            message = '{} ERROR: volume-id={} and tags={}\n{}'.format(
+                log_prefix, volume_id, self.tags, error)
+            self.logger.error(message)
+            if snapshot or snapshot_creation_operation:
+                self.delete_snapshot(snapshot_id)
+                snapshot = None
+            raise Exception(message)
+
+        return snapshot
+
+    def _get_snapshot_list(self, snapshot_id):
+        region_id = self.__aliCredentials['region_name']
+        get_snapshot_req_params = {
+            'PageSize' : 10,
+            'RegionId' : region_id,
+            'SnapshotIds' : [snapshot_id]
+        }
+        get_snapshot_request = self._get_common_request('DescribeSnapshots', get_snapshot_req_params)
+        snapshot_details = self.compute_client.do_action_with_exception(get_snapshot_request)
+        snapshot_details_json = json.loads(snapshot_details.decode('utf-8'))
+        return snapshot_details_json['Snapshots']['Snapshot']
+    
+    def _get_snapshot_with_exception(self, snapshot_id):
+        snapshot_list = self._get_snapshot_list(snapshot_id)
+        if len(snapshot_list) > 1:
+            message = 'More than 1 snapshot found with id {}'.format(
+            snapshot_id)
+            raise Exception(message)
+        if len(snapshot_list) == 0:
+            message = 'Snapshot with id {} is not found'.format(
+            snapshot_id)
+            raise Exception(message)
+        snapshot = snapshot_list[0]
+        return Snapshot(snapshot['SnapshotId'], snapshot['SourceDiskSize'], snapshot['CreationTime'], snapshot['Status'])
+
+    def get_snapshot(self, snapshot_id):
+        try:
+            return self._get_snapshot_with_exception(snapshot_id)
+        except Exception as error:
+            self.logger.error('[ALI] ERROR: Error in getting snapshot {}.\n{}'.format(
+                snapshot_id, error))
+            return None
+    
+    def _is_snapshot_ready(self, snapshot_id):
+        """Gets the snapshot state.
+        https://www.alibabacloud.com/help/doc-detail/25641.htm?spm=a2c63.p38356.a3.5.c880458dkimfOs#SnapshotType
+        Status can be "progressing" "accomplished" or "failed"
+        """
+        # Try until snapshot state is either "accomplished" or "failed"
+        # Returns True if multiple snapshots with same id are found
+        # Retries for failures also
```

Comment:
What is the reason behind returning `True` when multiple snapshots are found? Is there any specific reason why the exception [here](https://github.com/cloudfoundry-incubator/service-fabrik-backup-restore/pull/57/files#diff-fe5e9623c538d6846d10874ca9426104R219) is not raised as part of `_is_snapshot_ready ` itself?

Reply:
If we get multiple snapshots then there is no point in retrying for snapshot to get ready (We retry on exception). It will through exception in `create_snapshot` method [here](https://github.com/cloudfoundry-incubator/service-fabrik-backup-restore/pull/57/files#diff-fe5e9623c538d6846d10874ca9426104R181)

Reply:
Understood.